openassessment/xblock/submission_mixin.py

```diff
@@ -239,60 +243,143 @@ def get_saved_file_metadata(self):
+        return _safe_load_json_list(self.saved_files_names)
+
+    def set_file_names(self, file_name_list):
+        self.saved_files_names = json.dumps(file_name_list)
+
+    def get_file_sizes(self):
+        return _safe_load_json_list(self.saved_files_sizes)
+
+    def set_file_sizes(self, file_size_list):
+        self.saved_files_sizes = json.dumps(file_size_list)
+
+    def fix_file_names(self, descriptions=None):
+        descriptions = descriptions or self.get_file_descriptions()
+        if len(self.get_file_names()) != len(descriptions):
+            self.set_file_names([None for _ in range(len(descriptions))])
+        return self.get_file_names()
+
+    def fix_file_sizes(self, descriptions=None):
+        descriptions = descriptions or self.get_file_descriptions()
+        if len(self.get_file_sizes()) != len(descriptions):
+            self.set_file_sizes([None for _ in range(len(descriptions))])
+        return self.get_file_sizes()
+
+    def get_safe_normalized_file_metadata(self):
+        descriptions = self.get_file_descriptions()
+        names = self.fix_file_names(descriptions)
+        sizes = self.fix_file_sizes(descriptions)
+        return descriptions, names, sizes
+
+    def append_safe_normalized_file_metadata(self, descriptions_to_add, names_to_add, sizes_to_add):
+        """
+        Given lists of new file metadata, write the new metadata to our stored file metadata fields
+
+        Args:
+            descriptions_to_add: a list of file descriptions
+            names_to_add: a list of file names
+            sizes_to_add: a list of file sizes as integers
+
+        Returns: newly updated file metadata fields
+        """
+        if not (len(descriptions_to_add) == len(names_to_add) == len(sizes_to_add)):
+            message = (
+                'Attempted to append file metadata of differing lengths: '
+                'descriptions = {}, names = {}, sizes = {}'
+            )
+            raise FileUploadError(message.format(descriptions_to_add, names_to_add, sizes_to_add))
+
+        existing_file_descriptions, existing_file_names, existing_file_sizes = self.get_safe_normalized_file_metadata()
+
+        new_descriptions = existing_file_descriptions + descriptions_to_add
+        self.set_file_descriptions(new_descriptions)
+
+        new_names = existing_file_names + names_to_add
+        self.set_file_names(new_names)
+
+        new_sizes = existing_file_sizes + sizes_to_add
+        self.set_file_sizes(new_sizes)
+
+        return new_descriptions, new_names, new_sizes
+
+    def delete_safe_normalized_file_metadata(self, index):
+        """
+        Given a file index to remove, null out its metadata in our stored file metadata fields
+
+        Args:
+            index: file index to remove
+
+        Returns: newly updated file metadata fields
+        """
+        existing_file_descriptions, existing_file_names, existing_file_sizes = self.get_safe_normalized_file_metadata()
```

Comment:
Nit: `existing_file_descriptions` etc. is a bit of a misnomer at the end of this function, since we modify and return it.  Maybe `stored_file_descriptions` etc. instead?